flower/utils/results/stores/redis.py

```diff
@@ -0,0 +1,31 @@
+import json
+from collections.abc import Iterator
+from typing import Any
+
+from celery.backends.redis import RedisBackend
+
+from flower.utils.results.result import Result, ResultIdWithResultPair, ResultHeap
+from flower.utils.results.stores import AbstractBackendResultsStore
+
+
+class RedisBackendResultsStore(AbstractBackendResultsStore[RedisBackend]):
+    def results_by_timestamp(self, limit: int | None = None, reverse: bool = True) -> Iterator[
+        ResultIdWithResultPair
+    ]:
+        heap_size_limit = self.max_tasks_in_memory
+        if limit is not None and limit < heap_size_limit:
+            heap_size_limit = limit
+
+        task_key_prefix = self.backend.task_keyprefix
+
+        heap = ResultHeap(heap_size_limit=heap_size_limit, reverse_ordering=reverse)
+
+        for key in self.backend.client.scan_iter(
+            match=task_key_prefix + ("*" if isinstance(task_key_prefix, str) else b"*")
+        ):
+            result_data: dict[str, Any] = json.loads(self.backend.client.get(key))
+            result = Result(**result_data)
+            heap.push(result)
+
+        for task_id, result in heap:
+            yield task_id, result
```

Comment:
thoughts on first making a local full copy, then yielding item-wise?
I tend to like lazy code, i would expect to see more like a batch read (redis api is batch size must be one record at a time? that kinda stinks for a table scan) and then yielding item-wise from that batch, which in this case would probably be yield for each read

Reply:
I definitely want to avoid making a full in-memory copy of the Redis values. That would be extremely memory-intensive. The intent behind using a heap is that we can limit the number of results in memory at any given time, while still ending up with a properly-sorted result at the end. That said, if there is a more efficient way to "table scan" redis using batching, I'm all for it. I just used `scan_iter()` because that's the one way I know how to reliably iterate over a keyset. 

This will be much less relevant once we start implementing the SQLAlchemy results store. Unlike Redis, SQL allows us to be smart enough to encode sorting/filtering in the query itself, so we don't have to worry about table scans over there.

Reply:
oh help me understand `heap`, i thought it was implemented as a [list](https://github.com/shipwell/flower/pull/1/files/c6ea189a27772c49511b0f10325db8b958a2d612#diff-9f6b61eaf1c10c7bd0d867ccbff5d028aaa96abc1bfa79705417939074f29664R96), i don't see where we limit its size or do like an LRU eviction 

or rather, the implementation is to only store the (task_id, date_done.timestamp) tuple so its okay to scan the entire redis db into memory for just these two values.

I think i understand this, did i get it right?

Reply:
> or rather, the implementation is to only store the (task_id, date_done.timestamp) tuple so its okay to scan the entire redis db into memory for just these two values.

This is not the case; we do indeed store the entire `Result` object, which has fields like `args` and `kwargs` that may be much larger than just a (string, timestamp) pair. 

> oh help me understand heap, i thought it was implemented as a [list](https://github.com/shipwell/flower/pull/1/files/c6ea189a27772c49511b0f10325db8b958a2d612#diff-9f6b61eaf1c10c7bd0d867ccbff5d028aaa96abc1bfa79705417939074f29664R96), i don't see where we limit its size or do like an LRU eviction

Yes, I believe this is the source of the misunderstanding. The functions within the Python [heapq](https://docs.python.org/3/library/heapq.html) module all operate on a `list` data structure, and by default they don't enforce an upper bound on the size of that list. However, our `ResultHeap.push()` method checks the size of the list every time we attempt to push an element. If we are already at max capacity, we evict an element at the same time as we push the new one by using `heapq.heappushpop()`. This behavior is what allows us to manage memory more safely than we would by simply loading the entire Redis dataset into memory.